### PR TITLE
[ADR-68] Mutualize Cassandra storage for Quota

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+One generic threat is resource abuse. Meaning a user uses an unfair amount of resources.
+
+Similarily resource limitation is a way to decrease operating costs.
+
+Quota, standardized in IMAP via [RFC-9208](https://www.rfc-editor.org/rfc/rfc9208.html) and in JMAP via
+[RFC-9425](https://datatracker.ietf.org/doc/rfc9425/) are a mechanism designed to keep track of
+user current usage of the resource and match it against the stored limit.
+
+Resource limitation is a broad use case that we might want t better enforce in Apache James. This is especially
```
Comment (Arsnael) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:21]:
```suggestion
Resource limitation is a broad use case that we might want to better enforce in Apache James. This is especially
```

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+Port exising Cassandra implementations of quota for various services to use those helpers. Namely the Cassandra mailbox, Sieve storage. Use those
+helpers to implement JMAP quota uploads.
+
+## Consequences
+
+We expect a small decrease of the count of tables in our Cassandra schema, going from 7 (one for mailbox quota current value, one for mailbox quota
+per user limit, one for mailbox quota per domain limit, one for mailbox quota global limit, one for Sieve uota current value, one for Sieve quota
+per user limit, one for Sieve Quota global limit, one for JMAP uploads current value) to just 2 (helper tables: one for current value, one for limits).
+
+We expect to make it easier to implement quta related features in the future.
```
Comment (Arsnael) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:59]:
```suggestion
We expect to make it easier to implement quota related features in the future.
```

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+We need to set up Cassandra migration tasks and to record the schema upgrade into the upgrade instructions. A compaibility mode will need to be
+explicitly cnfigured in order to run the migration.
```
Comment (Arsnael) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:62]:
```suggestion
explicitly configured in order to run the migration.
```

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+## Alternatives
+
+Keep the current, exploded table structure and either do not care about table count increase or do not implement further impotant resource limitation
```
Comment (Arsnael) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:66]:
```suggestion
Keep the current, exploded table structure and either do not care about table count increase or do not implement further important resource limitation
```

**Review by Arsnael — APPROVED** [review]

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+## Decision
+
+Add Cassandra storage helpers, exposing a generic interface for Cassandra quota implementation of various services to mutualize their
+storage. This relies on a part of the primary key allowing to distinguish the component.
+
+![Architecture diagram for quota related features](img/adr-68-architecture-diagram.png)
+
+The helper will consist of 2 tables:
+- One composed of counters to track current value
+- One comosed of regular `long` to track limits
```
Comment (ottoka) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:46]: - One com**p**osed of regular `long` to track limits

File: src/adr/0068-mutualize-cassandra-storge-for-quota.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
```
Comment (chibenwa) [inline src/adr/0068-mutualize-cassandra-storge-for-quota.md:46]:
```suggestion
- One composed of regular `long` to track limits
```

(sorry keyboard issues)

**Pull request merged**